### ISSUE TITLE
weighted_similarity_chunk maybe could be optimized

### ISSUE THREAD

**Issue by chhotii-alex** [ArnaoutLab/diversity#84]
Which of these would you think would be faster?
Option 1: Create Python list of Python lists and pass to array constructor:
```
matrix  = np.array([
    [myfunc(i, j)
        for i in range(x)]
       for j in range(x)
])
```
Option 2: Create an np.array and set its values by addressing:
```
matrix = np.empty( shape=(x, x))
for i in range(x):
    for j in range(x):
        matrix[i, j] = func(i, j)
 ```
I would've thought "of course the second one would be WAY faster", StackOverflow seems to agree, and that's how SimilarityFromFunction is implemented. However, a wee bit of testing on my (Apple M1 Pro) laptop seems to suggest otherwise, and that making the Python lists is running about 40% faster. Requires further investigation, but suggests that we could be getting a speed-up by optimizing how array values are assigned.

**Comment by chhotii-alex** [comment]
Interestingly, using numpy's fromfunction array construction routine seems to be way faster. Way way faster. (Tested on Windows.) See https://numpy.org/doc/stable/reference/generated/numpy.fromfunction.html

**Comment by chhotii-alex** [comment]
Oh, no, `fromfunction` is not suitable, it isn't doing what I thought. Scratch that.

However, using a generator comprehension looks very promising.

**Comment by chhotii-alex** [comment]
Nope, scratch that, trying to use a generator comprehension also does not work.

**Comment by chhotii-alex** [comment]
Something like this may be optimal:
```
def from_iter():
    my_iter = (
        myfunc(j, i) for i in range(x) for j in range(x)
    )
    matrix = np.fromiter(my_iter, float, count=(x*x))
    return matrix.reshape( (x, x) )
```

**Comment by chhotii-alex** [comment]
This issue arose because I was seeking to verify my claim that passing in a similarity function (and thus using Ray under  the covers) should  result in up to an N-fold speed-up, where N was the number of cores. I observed that a given job, using the similarity function, took about 90 minutes on a 10-core machine. I would then expect that making the similarity matrix in a single thread would take  not much less than 15 hours; however tqdm was predicting about  11 hours. If the single-threaded  similarity matrix creation was 11 hours then why would the parallelized version take about 36% more compute time?

I don't know how much intrinsic overhead is involved in using Ray. Do the workers share memory, or is there a lot of copying from one memory space to another? I suppose if the anticipated typical use case is to distribute jobs across multiple machines, enabling shared memory has been a low priority. Also, there's the overhead of coordination. Still, I'm surprised that all that imposes a 36% compute time tax.

But it's a bit of an apples-to-oranges comparison. My single-threaded code calls the array constructor on the Python list of lists, which to my surprise is faster. Also, looking at tqdm progress did not take into account things that happen along the way when parallelized, but only after tqdm iteration single-threaded-- conversion to np.array, and the actual matrix multiplication. Need to do a more fair head-to-head comparison.

I think that they method of assigning to the np.array is only a miniscule, insignificant factor here. Moving to a generator function is perhaps premature optimization, given that this wouldmake the code in `similarity.py` more complicated and less readable. 

So I'm making all these  notes on what I'm thinking because this issue is being demoted to very low priority.
